fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinLAsscomHelper.java

```diff
@@ -0,0 +1,219 @@
+
+        Preconditions.checkArgument(topJoin.getCondition().isPresent(), "topJoin onClause must be present.");
+        topJoinOnClause = (Expression) topJoin.getCondition().get();
+        Preconditions.checkArgument(bottomJoin.getCondition().isPresent(), "bottomJoin onClause must be present.");
+        bottomJoinOnClause = (Expression) bottomJoin.getCondition().get();
+
+        aOutputSlots = Utils.getOutputSlotReference(a);
+        bOutputSlots = Utils.getOutputSlotReference(b);
+        cOutputSlots = Utils.getOutputSlotReference(c);
+    }
+
+    public static JoinLAsscomHelper of(LogicalJoin topJoin, LogicalJoin<GroupPlan, GroupPlan> bottomJoin) {
+        return new JoinLAsscomHelper(topJoin, bottomJoin);
+    }
+
+    /**
+     * Get the onCondition of newTopJoin and newBottomJoin.
+     */
+    public boolean initJoinOnCondition() {
+        List<Expression> topJoinOnClauseConjuncts = ExpressionUtils.extractConjunction(topJoinOnClause);
+        for (Expression topJoinOnClauseConjunct : topJoinOnClauseConjuncts) {
+            // Ignore join with some OnClause like:
+            // Join C = B + A for above example.
+            if (ExpressionUtils.isIntersecting(
+                    topJoinOnClauseConjunct.collect(SlotReference.class::isInstance), aOutputSlots)
+                    && ExpressionUtils.isIntersecting(
+                    topJoinOnClauseConjunct.collect(SlotReference.class::isInstance), bOutputSlots)
+                    && ExpressionUtils.isIntersecting(
+                    topJoinOnClauseConjunct.collect(SlotReference.class::isInstance), cOutputSlots)
+            ) {
+                return false;
+            }
+        }
+        List<Expression> bottomJoinOnClauseConjuncts = ExpressionUtils.extractConjunction(
+                bottomJoinOnClause);
+
+        List<Expression> allOnCondition = Lists.newArrayList();
+        allOnCondition.addAll(topJoinOnClauseConjuncts);
+        allOnCondition.addAll(bottomJoinOnClauseConjuncts);
+
+        List<SlotReference> newBottomJoinSlots = Lists.newArrayList();
+        newBottomJoinSlots.addAll(aOutputSlots);
+        newBottomJoinSlots.addAll(cOutputSlots);
+
+        for (Expression onCondition : allOnCondition) {
+            List<SlotReference> slots = onCondition.collect(SlotReference.class::isInstance);
+            if (new HashSet<>(newBottomJoinSlots).containsAll(slots)) {
+                newBottomJoinOnCondition.add(onCondition);
+            } else {
+                newTopJoinOnCondition.add(onCondition);
+            }
+        }
+
+        // newBottomJoinOnCondition/newTopJoinOnCondition is empty. They are cross join.
+        // Example:
+        // A: col1, col2. B: col2, col3. C: col3, col4
+        // (A & B on A.col2=B.col2) & C on B.col3=C.col3.
+        // (A & B) & C -> (A & C) & B.
+        // (A & C) will be cross join (newBottomJoinOnCondition is empty)
+        if (newBottomJoinOnCondition.isEmpty() || newTopJoinOnCondition.isEmpty()) {
```

Comment:
we need to do infer predicate later to process
`a join b on a.id = b.id join c on b.id = c.id`
it could be transform to
`a join c on a.id = c.id join b on b.id = c.id`